docs/proposals/20220221-runtime-SDK.md

```diff
@@ -0,0 +1,883 @@
+
+## Runtime Extensions developer guide
+
+As a developer building systems on top of Cluster API, if you want to hook in the Cluster’s lifecycle via
+a Runtime Hook, you are required to implement a Runtime Extension handling requests according to the
+OpenAPI specification for the Runtime Hook you are interested in.
+
+Runtime Extensions by design are very powerful and flexible, however given that with great power comes
+great responsibility, a few key consideration should always be kept in mind (more details in the following paragraphs):
+
+- Runtime Extension are components that should be designed, written and deployed with great caution given that they
+  can affect the proper functioning of the Cluster API runtime.
+- Cluster administrators should carefully vet any Runtime Extension registration, thus preventing malicious components
+  from being added to the system.
+
+Please note that following similar practices is already commonly accepted in the Kubernetes ecosystem for
+Kubernetes API server admission webhooks, and Runtime Extensions share the same foundation and most of the same
+considerations/concerns apply.
+
+### Implementing Runtime Extensions
+
+As a developer building systems on top of Cluster API, if you want to hook in the Cluster’s lifecycle via a
+Runtime Extension, you are required to implement an HTTP server handling a discovery request and a set of 
+additional requests according to the OpenAPI specification for the Runtime Hook you are interested in.
+
+E.g.
+
+```go
+// Note: this is pseudo code, meant to demonstrate that implementing a RuntimeExtension requires only minimal scaffolding;
+// the exact details will be defined during implementation, possibly taking advantage of Golang generics.
+
+var c = catalog.NewCatalog()
+
+func init() {
+    v1alpha2.AddToCatalog(c)
+}
+
+func main() {
+    ctx := context.Background()
+
+    listener, err := net.Listen("tcp", net.JoinHostPort("127.0.0.1", "8082"))
+    if err != nil {
+        panic(err)
+    }
+
+    // Build an HTTP handler for a given operation, calling a strongly typed func at each request
+    beforeUgradeHandler, err := catalogHTTP.NewHandlerBuilder().
+        WithCatalog(c).
+        AddService(&v1alpha2.DiscoveryHook{}, doDiscovery).
+        AddService(&v1alpha2.BeforeUgradeHook{}, doBeforeUpgrade).
+        Build()
+    if err != nil {
+        panic(err)
+    }
+
+    srv := &http.Server{
+        Handler: BeforeUgradeHandler,
+    }
+    if err := srv.Serve(listener); err != nil && err != http.ErrServerClosed {
+        panic(err)
+    }
+}
+
+func doBeforeUpgrade(in *v1alpha2.DiscoveryInput, out *v1alpha2.DiscoveryOutput) error {
+    out.Items = append(out.Items,
+        *v1alpha2.DiscoveryExtension{
+            name: "upgradeAddons",
+            hook: *v1alpha2.DiscoveryHook{
+                apiVersion: "hook.runtime.cluster.x-k8s.io/v1alpha1",
+                name: "BeforeUpgrade",
+            },
+            timeoutSeconds: 5,
```

Comment:
Would it be useful to have priority for a hook, for cases where some kind of ordering might be required to co-ordinate between hooks of the same type?

Reply:
In this iteration of the proposal we would like to avoid introducing dependencies like this:

> Instead, unless there's a strong and evident need for it, we are not considering adding support for defining dependencies among Runtime Extensions, being it modeled with something similar to [systemd unit options](https://www.freedesktop.org/software/systemd/man/systemd.unit.html) or alternative approaches.

> The main reason behind that is that such type of feature introduces complexity and creates "pet" like relations across components making the overall system more fragile. This is also consistent with the [avoid dependencies](https://github.com/kubernetes-sigs/cluster-api/blob/9a3a8f557db850b4a67e9e8e0a70cb2d0261caee/docs/proposals/20220221-runtime-SDK.md#avoid-dependencies) recommendation above.

Reply:
As explained by @sbueringer we are not going to support dependencies between extensions in the first iteration.